docs/concepts/unions.md

```diff
@@ -78,23 +78,62 @@ print(User(id='456'))  # (2)
 
 ### Smart Mode
 
-Because of the surprising side effects of `union_mode='left_to_right'`, in Pydantic >=2 the default mode for `Union` validation is `union_mode='smart'`.
+Because of the potentially surprising results of `union_mode='left_to_right'`, in Pydantic >=2 the default mode for `Union` validation is `union_mode='smart'`.
 
-In this mode, pydantic scores a match of a union member into one of the following three groups (from highest score to lowest score):
+In this mode, pydantic attempts to select the best match for the input from the union members. The exact algorithm may change between Pydantic minor releases to allow for improvements in both performance and accuracy.
+
+!!! tip
+
+    In general, we recommend using [discriminated unions](#discriminated-unions). They are both more performant and more predictable than untagged unions, as they allow you to control which member of the union to validate against.
+
+    For complex cases, if you're using untagged unions, it's recommended to use `union_mode='left_to_right'` if you need guarantees about the order of validation attempts against the union members.
+
+    If you're looking for incredibly specialized behavior, you can use a [custom validator](../concepts/validators.md#field-validators).
+
+1. The number of valid fields set (relevant for models, dataclasses, and typed dicts)
+2. The exactness of the match (relevant for all types)
+
+#### Number of valid fields set
+
+!!! note
+    This metric was introduced in Pydantic v2.8.0. Prior to this version, only exactness was used to determine the best match.
+
+This metric is currently only relevant for models, dataclasses, and typed dicts.
+
+The greater the number of valid fields set, the better the match. The number of fields set on nested models is also taken into account.
+These counts bubble up to the top-level union, where the union member with the highest count is considered the best match.
+
+For data types where this metric is relevant, we prioritize this count over exactness. For all other types, we use solely exactness.
+
+#### Exactness
+
+For `exactness`, Pydantic scores a match of a union member into one of the following three groups (from highest score to lowest score):
 
 - An exact type match, for example an `int` input to a `float | int` union validation is an exact type match for the `int` member
 - Validation would have succeeded in [`strict` mode](../concepts/strict_mode.md)
 - Validation would have succeeded in lax mode
 
-The union match which produced the highest score will be selected as the best match.
+The union match which produced the highest exactness score will be considered the best match.
```

Comment:
I still think all of lines 93-116 should be moved into the "Smart Mode Algorithm" box.